Automattic/wp-calypso
# Do not transpile JS packages twice

### Summary

Do not transpile some packages at installation time because webpack/jest will transpile them anyway. This allows faster installations (~38% faster) and smaller bundles.

Based on #43679

### Background

Packages in `./packages/` implement a `prepare` script that is run every time we do `yarn install`. For JavaScript packages, this script is using babel to transpile the source to CJS and ESM. Those packages expose the ESM result in `pkg.module` (used by Webpack) and the CJS result in `pkg.main` (used by Node). 

Later, when compiling `./client`, webpack loads the ESM transpiled files, transpiles them again and add them to the bundles.

### Changes

* Do not implement a `prepare` script. Instead, use the `prepack` script to transpile to CJS. This will run the transpilation when `yarn publish` or `yarn pack`, but not on installation (see https://docs.npmjs.com/misc/scripts)

* Change `module` to point to the source code (not the transpiled ESM). This allows distributing the untranspiled source code so consumers of this package (including Calypso) have control over the transpilation options. Specifically, this allows Calypso to transpile them for "evergreen" browsers. Before this change, packages in `./packages` were always transpiled for "fallback" browsers.

* Drop the `prepare` script. This will allow faster `yarn build-packages` (which is run on every `yarn install`). In my machine, in master `yarn build-packages` takes 39.9s to "prepare" 23 packages. After this change, it takes 24.6s (**38% faster**) to prepare 10 packages.

### Testing instructions

* Verify calypso.live loads (both evergreen and fallback) and there are no noticeable errors. If there is something broken entire sections will fail to load, so it should be easy to spot.
* Verify that ICFY reports no size changes (so no new packages are added to the bundle).
* Verify that tests passes (so Jest also understand the changes).

### Update (22/Oct)

- Instead of changing the meaning of `module`, this PR introduces a new `calypso:src` field that points to the untranspiled source code. `pkg.module` and `pkg.main` are unchanged.
